jmeter/src/main/java/org/eclipse/hono/jmeter/client/HonoCommander.java

```diff
@@ -0,0 +1,259 @@
+/*
+ * ******************************************************************************
+ *  * Copyright (c) 2016, 2018 Contributors to the Eclipse Foundation
+ *  *
+ *  * See the NOTICE file(s) distributed with this work for additional
+ *  * information regarding copyright ownership.
+ *  *
+ *  * This program and the accompanying materials are made available under the
+ *  * terms of the Eclipse Public License 2.0 which is available at
+ *  * http://www.eclipse.org/legal/epl-2.0
+ *  *
+ *  * SPDX-License-Identifier: EPL-2.0
+ *  ******************************************************************************
+ *
+ */
+
+package org.eclipse.hono.jmeter.client;
+
+import static org.eclipse.hono.util.MessageTap.getConsumer;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+
+import org.apache.jmeter.samplers.SampleResult;
+import org.apache.qpid.proton.amqp.messaging.Data;
+import org.apache.qpid.proton.message.Message;
+import org.eclipse.hono.client.CommandClient;
+import org.eclipse.hono.client.HonoClient;
+import org.eclipse.hono.client.MessageConsumer;
+import org.eclipse.hono.client.impl.HonoClientImpl;
+import org.eclipse.hono.config.ClientConfigProperties;
+import org.eclipse.hono.jmeter.HonoCommanderSampler;
+import org.eclipse.hono.jmeter.sampleresult.HonoCommanderSampleResult;
+import org.eclipse.hono.util.TimeUntilDisconnectNotification;
+import org.slf4j.LoggerFactory;
+
+import io.vertx.core.Future;
+import io.vertx.core.buffer.Buffer;
+
+/**
+ * Hono commander, which creates command client, send commands to devices and receive responses from devices;
+ * asynchronous API needs to be used synchronous for JMeters threading model.
+ */
+public class HonoCommander extends AbstractClient {
+
+    private static final org.slf4j.Logger LOG = LoggerFactory.getLogger(HonoCommander.class);
+    private final HonoClient client;
+    private final HonoCommanderSampler sampler;
+    private final List<String> devicesReadyToReceiveCommands = new CopyOnWriteArrayList<>();
+    private final AtomicInteger successCount = new AtomicInteger(0);
+    private final AtomicInteger errorCount = new AtomicInteger(0);
+    private final AtomicLong bytesSent = new AtomicLong(0);
+    private final AtomicLong bytesReceived = new AtomicLong(0);
+    private final AtomicLong sampleStart = new AtomicLong(0);
```

Comment:
I don't see what the Atomic's helping with. FMPOV you need to to be sure, that the sequence of writing the values to the sample and then set them to zero at each call of sample is synchronized against the increments of the values from the command receiving. If you would just use Atomics, you stay with a possible leak between setting in sample and resetting to zero. Und if you use synchronized for this sequence you need the same synchronization lock for the increments.

Reply:
During every sampling , sampler resets those counts to 0 and on the other side vertx threads set them to actual value on success or failure. In order to ensure writing by these threads, Atomic is used. When vertx threads set these count, synchronized locks were used.

Reply:
You use synchronized on the same lock object in the sample (JMeter thread) and in the command response. You do not use it for errorCount. I think this is wrong. The errorCount could be incremented while the sample method is invoked by the JMeter thread. And this could happen between `result.setErrorCount(errorCount.get());` and `errorCount.set(0);` since the atomic does not help here. In this case one error would be gone.

Reply:
Thank you. I got your point. Yes, I see now the chance of counts getting missed out. I would like to replace `result.setErrorCount(errorCount.get())` and `errorCount.set(0)` with `result.setErrorCount(errorCount.getAndSet(0))`. With this, it can be ensured that error counts won't be missed. Also with this way, I think we can avoid `synchronized(lock)`block inside `sample()` method. What do you say?

Reply:
There are still the two operations `errorCount.getAndSet(0)` and `result.setErrorCount(..)` as a sequence in which the `errorCount.set(0)` can happen. I do not see how the synchronized could be avoided.

Reply:
getAndSet(0) is atomic, it returns the value and the set the value to 0. Hence errorCount.set(0) is no more needed.

Reply:
Sorry, I meant `errorCount.addAndGet(1);`, not `errorCount.set(0)`. 

`getAndSet(0)` is atomic. This is called.. and after it `result.setErrorCount()` with the value from `getAndSet(0)` is called. And in between, `errorCount.addAndGet(1);` could be called from the vert.x thread.